File: Arrays/CustomArray.java
/**
 * Simple implementation of a dynamic array
 */
class MyArray<T> {
    private Integer currentIndex;
    private Integer capacity;
    private T data[];

    public MyArray() {
        this.capacity = 1; // initial value
        this.currentIndex = 0;
        data = new T[this.capacity]; // fix this

    }

    public void push(T item) {
        if( currentIndex == capacity) {
            T tempCopy[] = data;
            capacity *=2; // double capacity;
            data = new T[this.capacity]; // fix this
        }
        data[currentIndex++] = item;
    }

    public T pop() {
        return data[currentIndex--];
    }

    public T get(Integer index) {
        if(index < 0 || index >= currentIndex)
            return null;

        return data[index];
    }

    public Integer length() {
        return currentIndex;
    }


}
public class CustomArray {
    public static void main(String args[]) {

        MyArray<Integer> myArray = new MyArray<Integer>();
        myArray.push(2);
        myArray.push(4);
        myArray.push(6);
        System.out.println("Array Length is: " + myArray.length());
        System.out.println("Array[2] = " + myArray.get(2));
        myArray.pop();
        System.out.println("Length after poping is: " + myArray.length());
        
    }
}

File: Arrays/CustomArray.cpp
#include <iostream>
#include <memory>
#include <string>
template <typename T>
class MyArray
{
private:
    int currentIndex;
    int capacity;
    std::unique_ptr<T[]> data;

public:
    MyArray() : currentIndex(0), capacity(1)
    {
        data = std::make_unique<T[]>(capacity);
    }
    void push(T item)
    {
        if (currentIndex == capacity)
        {
            auto tempCopy = std::move(data); // copy prev content

            capacity *= 2;                          // double the capacity
            data = std::make_unique<T[]>(capacity); // get some brand new memory :)
            for (int i = 0; i < currentIndex; i++)
            {
                data[i] = tempCopy[i]; // copy the prev content back to our new memory, not so new now :(
            }
        }
        data[currentIndex++] = item;
    }

    T pop()
    {
        return data[currentIndex--]; // not really pop, just a dirty hack :)
    }
    T get(int index)
    {
        if (index < 0 || index >= currentIndex)
            return NULL; // again a hack, will throw exeption with std::string etc
        return data[index];
    }
    int length()
    {
        return currentIndex;
    }
};

// Example usage
int main()
{
    MyArray<int> myArray;
    myArray.push(1);
    myArray.push(3);
    myArray.push(5);
    std::cout << myArray.length() << std::endl;
    std::cout << myArray.get(2) << std::endl;
    myArray.pop();
    std::cout << myArray.get(2) << std::endl;
    std::cout << myArray.length() << std::endl;
}